native/src/render_target_stack.rs

```diff
@@ -0,0 +1,133 @@
+
+    fn last_token_is_a_newline(&self) -> bool {
+        match self {
+            Self::BaseQueue(ltc) => ltc.last_token_is_a_newline(),
+            Self::BreakableEntry(be) => be.last_token_is_a_newline(),
+        }
+    }
+
+    fn insert_comments_at_last_hard_newline(&mut self, comments: CommentBlock) {
+        match self {
+            RenderTarget::BaseQueue(rq) => rq.insert_comments_at_last_hard_newline(comments),
+            RenderTarget::BreakableEntry(be) => be.insert_comments_at_last_hard_newline(comments),
+        }
+    }
+
+    fn push_token(&mut self, lt: LineToken) {
+        match self {
+            RenderTarget::BaseQueue(rq) => rq.push(lt),
+            RenderTarget::BreakableEntry(be) => be.push(lt),
+        }
+    }
+}
+
+pub struct RenderTargetStack {
+    target_stack: Vec<RenderTarget>,
+}
+
+impl RenderTargetStack {
+    pub fn new() -> Self {
+        RenderTargetStack {
+            target_stack: vec![RenderTarget::BaseQueue(LineTokenCollection::new())],
+        }
+    }
+
+    pub fn push_token(&mut self, lt: LineToken) {
+        self.current_target_mut().push_token(lt);
+    }
+
+    pub fn insert_comments_at_last_hard_newline(&mut self, comments: CommentBlock) {
+        self.current_target_mut()
+            .insert_comments_at_last_hard_newline(comments);
+    }
+
+    pub fn last_token_is_a_newline(&self) -> bool {
+        self.current_target().last_token_is_a_newline()
+    }
+
+    pub fn insert_extra_newline_at_last_newline(&mut self) {
+        self.current_target_mut()
+            .insert_extra_newline_at_last_newline();
+    }
+
+    pub fn push_breakable_entry(&mut self, be: BreakableEntry) {
+        self.target_stack.push(RenderTarget::BreakableEntry(be));
+    }
+
+    pub fn pop_expecting_breakable_entry(&mut self) -> BreakableEntry {
+        if let RenderTarget::BreakableEntry(be) =
+            self.target_stack.pop().expect("should be present")
+        {
+            return be;
+        }
+        panic!("should be impossible")
+    }
+
+    pub fn record_line_number(&mut self, ln: LineNumber) {
+        for item in self.target_stack.iter_mut() {
+            if let RenderTarget::BreakableEntry(be) = item {
+                be.push_line_number(ln)
+            }
+        }
+    }
+
+    pub fn current_breakable_is_multiline(&self) -> Option<bool> {
+        match self.current_target() {
+            RenderTarget::BreakableEntry(be) => Some(be.is_multiline()),
+            _ => None,
+        }
+    }
+
+    pub fn into_render_queue(self) -> LineTokenCollection {
+        if self.target_stack.len() != 1 {
+            panic!("target stack was consumed when not length 1, indicates programming error");
+        }
+
+        let mut ts = self.target_stack;
+
+        match ts.pop().expect("we checked it exists") {
+            RenderTarget::BaseQueue(x) => x,
+            RenderTarget::BreakableEntry(_) => {
+                panic!("should not have breakable entry at bottom of stack")
```

Comment:
Is there any way we can use the type system to enforce these invariants? Should an issue be opened if they're violated?